### PR TITLE
Valeria/bubble correction

### PR DESCRIPTION
**Pull request by valeriabarra**
This is a ~~(WIP for now)~~ PR that addresses #1020 and #1019. 

It implements the bubble correction a la TempestRemap and HOMME. 
 

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Comment by valeriabarra** [comment]
bors try

**Comment by valeriabarra** [comment]
@charleskawczynski I decided to change the assert with a warning and do not perform the bubble correction algorithm at all in that case (similar to TempestRemap's check). This is the only clean way to deal with this in the case of low precision tests. I verified that this works as well in the ClimaAtmos case that was failing w/ Float32 earlier

**Comment by charleskawczynski** [comment]
> @charleskawczynski I decided to change the assert with a warning and do not perform the bubble correction algorithm at all in that case (similar to TempestRemap's check). This is the only clean way to deal with this in the case of low precision tests. I verified that this works as well in the ClimaAtmos case that was failing w/ Float32 earlier

ClimaAtmos uses Float32 by default, are you saying that the bubble correction won't work with Float32? I'm thinking that we may need feedback about this. My impression was that we could use this bubble correction with either precision (given the default precision in ClimaAtmos).

**Comment by valeriabarra** [comment]
> > @charleskawczynski I decided to change the assert with a warning and do not perform the bubble correction algorithm at all in that case (similar to TempestRemap's check). This is the only clean way to deal with this in the case of low precision tests. I verified that this works as well in the ClimaAtmos case that was failing w/ Float32 earlier
> 
> ClimaAtmos uses Float32 by default, are you saying that the bubble correction won't work with Float32? I'm thinking that we may need feedback about this. My impression was that we could use this bubble correction with either precision (given the default precision in ClimaAtmos).

No, I am not saying that. In lower precision test cases, only a few elements (I believe those that are flattest, near the center of the cube panels, not near the corners) won't "qualify" for bubble correction, meaning, it would be pointless and wasteful to apply it to them (this is similar to TempestRemap's [behavior](https://github.com/ClimateGlobalChange/tempestremap/blob/ddee889eacf4635c19abb3468bdc5d884e2c190c/src/FiniteElementTools.cpp#L377), that in that case does not apply the correction at all). This is because in lower precision test cases, the difference between the high-order element area and regular element area is not representable in that precision. So even correcting that, won't have any effect because literally in those cases we can't represent any better than that.

File: src/Spaces/spectralelement.jl [hunk 1]
```diff
@@ -211,20 +242,138 @@ function SpectralElementSpace2D(topology, quadrature_style)
...
+                    u, ∂u∂ξ = compute_local_geometry(
+                        global_geometry,
+                        topology,
+                        elem,
+                        ξ,
+                        AIdx,
+                    )
+                    J = det(Geometry.components(∂u∂ξ))
+                    WJ = J * quad_weights[i] * quad_weights[j]
+                    elem_area += WJ
```
Comment (charleskawczynski) [inline src/Spaces/spectralelement.jl:297]:
```suggestion
```

**Comment by charleskawczynski** [comment]
> ...In lower precision test cases, only a few elements (I believe those that are flattest, near the center of the cube panels, not near the corners) won't "qualify" for bubble correction, meaning, it would be pointless and wasteful to apply it to them (this is similar to TempestRemap's [behavior](https://github.com/ClimateGlobalChange/tempestremap/blob/ddee889eacf4635c19abb3468bdc5d884e2c190c/src/FiniteElementTools.cpp#L377), that in that case does not apply the correction at all). This is because in lower precision test cases, the difference between the high-order element area and regular element area is not representable in that precision. So even correcting that, won't have any effect because literally in those cases we can't represent any better than that.

Ah, ok, that makes sense given that these assertions are quite small. Sounds good to me. I've added one more comment about deleting an unused variable, aside from that I think we can rebase / fix the merge conflict and merge. Thank you!

**Review by charleskawczynski — APPROVED** [review]

File: test/Spaces/sphere.jl [hunk 2]
```diff
@@ -5,36 +5,51 @@ import ClimaCore: Domains, Topologies, Meshes, Spaces, Geometry, column
...
+        topology = Topologies.Topology2D(context, mesh)
+        quad = Spaces.Quadratures.GLL{Nq}()
+        space = Spaces.SpectralElementSpace2D(topology, quad)
+
+        # surface area
+        @test sum(ones(space)) ≈ FT(4pi * radius^2) rtol = 1e11 * eps(FT)
+
+        enable_bubble = false
+        no_bubble_space =
+            Spaces.SpectralElementSpace2D(topology, quad; enable_bubble)
```
Comment (simonbyrne) [inline test/Spaces/sphere.jl:24]:
```suggestion
        no_bubble_space =
            Spaces.SpectralElementSpace2D(topology, quad; enable_bubble = false)
```

File: test/Spaces/sphere.jl [hunk 3]
```diff
@@ -5,36 +5,51 @@ import ClimaCore: Domains, Topologies, Meshes, Spaces, Geometry, column
...
+        # Now check that constructor with enable_buble = false falls back on existing behavior
+        @test sum(ones(no_bubble_space)) ≈ FT(4pi * radius^2) rtol =
+            1e11 * eps(FT)
 
-    # surface area
-    @test sum(ones(space)) ≈ 4pi * radius^2 rtol = 1e-3
+        # Now check constructor with bubble enabled
+        enable_bubble = true
+        bubble_space =
+            Spaces.SpectralElementSpace2D(topology, quad; enable_bubble)
```
Comment (simonbyrne) [inline test/Spaces/sphere.jl:33]:
```suggestion
        bubble_space =
            Spaces.SpectralElementSpace2D(topology, quad; enable_bubble = true)
```

File: test/Spaces/sphere.jl [hunk 4]
```diff
@@ -5,36 +5,51 @@ import ClimaCore: Domains, Topologies, Meshes, Spaces, Geometry, column
...
-    domain = Domains.SphereDomain(radius)
-    mesh = Meshes.EquiangularCubedSphere(domain, ne)
-    topology = Topologies.Topology2D(context, mesh)
-    quad = Spaces.Quadratures.GLL{Nq}()
-    space = Spaces.SpectralElementSpace2D(topology, quad)
+    for FT in (Float64, Float32)
+        context = ClimaComms.SingletonCommsContext()
+        radius = FT(3)
+        ne = 4
+        Nq = 4
```
Comment (simonbyrne) [inline test/Spaces/sphere.jl:12]: Add a test for Nq = 2 as well?

File: test/Spaces/sphere.jl (hunk 2 above)
Comment (valeriabarra) [reply test/Spaces/sphere.jl:24]: Sorry, I am just seeing your comments now, after this has been merged.

File: src/Spaces/spectralelement.jl [hunk 5]
```diff
@@ -211,20 +242,137 @@ function SpectralElementSpace2D(topology, quadrature_style)
...
+                            elem,
+                            ξ,
+                            AIdx,
+                        )
+                        J = det(Geometry.components(∂u∂ξ))
+                        WJ = J * quad_weights[i] * quad_weights[j]
+                        # Modify WJ only for interior nodes
+                        if i != 1 && j != 1 && i != Nq && j != Nq
+                            WJ *= 1 + interior_elem_area
+                        end
```
Comment (simonbyrne) [inline src/Spaces/spectralelement.jl:366]:
```suggestion
                        # Modify J only for interior nodes
                        if i != 1 && j != 1 && i != Nq && j != Nq
                            J *= 1 + interior_elem_area
                        end
                        WJ = J * quad_weights[i] * quad_weights[j]
```

**Pull request merged**